x-pack/plugins/ml/public/data_frame/components/source_index_preview/source_index_preview.tsx

```diff
@@ -126,35 +124,65 @@ export const SourceIndexPreview: React.SFC<Props> = React.memo(({ cellClick, que
     setItemIdToExpandedRowMap({ ...itemIdToExpandedRowMap });
   }
 
-  useEffect(
-    () => {
-      setLoading(true);
-
-      ml.esSearch({
-        index: indexPattern.title,
-        rest_total_hits_as_int: true,
-        size: SEARCH_SIZE,
-        body: { query },
-      })
-        .then((resp: SearchResponse<any>) => {
-          const docs = resp.hits.hits;
-
-          if (selectedFields.length === 0) {
-            const newSelectedFields = getDefaultSelectableFields(docs);
-            setSelectedFields(newSelectedFields);
-          }
-
-          setTableItems(docs as EsDoc[]);
-          setLoading(false);
-        })
-        .catch((resp: any) => {
-          setTableItems([] as EsDoc[]);
-          setLoading(false);
-        });
-    },
-    [indexPattern.title, query.query_string.query]
+  const { errorMessage, status, tableItems } = useSourceIndexData(
```

Comment:
So instead of relying on useState is this just relying on it running on every render?

Reply:
It's a [custom hook](https://reactjs.org/docs/hooks-custom.html) which allows the reuse of state logic, so it relies on `useState` and other hooks internally https://github.com/elastic/kibana/pull/35490/files#diff-f9f36c0107448647d94ab7503e0ff5f0R32